django/thunderstore/repository/models/cache.py

```diff
@@ -111,3 +117,184 @@ def drop_stale_cache(cls):
+                content_encoding="gzip",
+            )
+
+        for listing in get_package_listings(community):
+            listing_bytes = listing_to_json(listing)
+
+            # Always add the first listing regardless of the size limit.
+            if not chunk_content:
+                chunk_content.extend(listing_bytes)
+            # Start new blob if adding current chunck would exceed the size limit.
+            # +2 for opening and closing brackets
+            elif len(chunk_content) + len(listing_bytes) + 2 > uncompressed_blob_size:
+                finalize_blob()
+                chunk_content = bytearray(listing_bytes)
+            else:
+                chunk_content.extend(b"," + listing_bytes)
+
+        if len(chunk_content) or not group.entries.exists():
+            finalize_blob()
+
+        group.set_complete()
+        index = get_index_blob(group)
+        cls.objects.create(community=community, index=index, chunks=group)
+
+    @classmethod
+    def drop_stale_cache(cls) -> None:
+        """
+        Delete objects from database and blob files from S3 buckets.
+        Cutoff period is used to ensure blobs still referenced by
+        cached data is not dropped prematurely.
+
+        TODO: only soft deletes the parent object until we've figured out
+        how to safely delete the blobs from main and mirror storages.
+
+        TODO: ensure dropping of stale caches doesn't fail if any blobs,
+        identified by hash of their content, is still used elsewhere.
+        """
+        for community in Community.objects.iterator():
+            latest = cls.get_latest_for_community(community)
+            if latest is None:
+                continue
+
+            cutoff = latest.created_at - timedelta(hours=cls.CACHE_CUTOFF_HOURS)
+            cls.objects.filter(created_at__lte=cutoff, community=community).update(
+                is_deleted=True,
+            )
+
+    @classmethod
+    def get_blob_content(cls, blob: DataBlob) -> List[Any]:
+        """
+        QoL method for returning the content of either index or chunk blob.
+        """
+        with gzip.open(blob.data, "rb") as f:
+            return json.loads(f.read())
+
+
+def get_package_listings(community: Community) -> models.QuerySet["PackageListing"]:
+    listing_ids = get_package_listing_queryset(community.identifier).values_list(
+        "id",
+        flat=True,
+    )
+    listing_ref = PackageListing.objects.filter(pk=models.OuterRef("pk"))
+
+    return order_package_listing_queryset(
+        PackageListing.objects.filter(id__in=listing_ids)
+        .select_related("community", "package", "package__owner")
+        .prefetch_related("categories", "community__sites", "package__versions")
+        .annotate(
+            _rating_score=models.Subquery(
+                listing_ref.annotate(
+                    ratings=models.Count("package__package_ratings"),
+                ).values("ratings"),
+            ),
+        ),
+    )
```

Comment:
This could be structured to be an enumerator that iteraters over the packages in chunks

Reply:
@MythicManiac Added a new commit https://github.com/thunderstore-io/Thunderstore/pull/1035/commits/b97e83dcbf8f1580f62e2715b76889f58567f8c4 for this.